sway-core/src/ir_generation/function.rs

```diff
@@ -2430,114 +2462,143 @@ impl<'eng> FnCompiler<'eng> {
+                    let mut cur_type_id = *base_type;
+                    // TODO-IG: Add support for projections being references themselves.
+                    for idx_kind in indices.iter() {
+                        let cur_type_info_arc = self.engines.te().get_unaliased(cur_type_id);
+                        let cur_type_info = &*cur_type_info_arc;
+                        match (idx_kind, cur_type_info) {
+                            (
+                                ProjectionKind::StructField { name: idx_name },
+                                TypeInfo::Struct(decl_ref),
+                            ) => {
+                                let struct_decl = self.engines.de().get_struct(decl_ref);
+
+                                match struct_decl.get_field_index_and_type(idx_name) {
+                                    None => {
+                                        return Err(CompileError::InternalOwned(
+                                            format!(
+                                            "Unknown field name \"{idx_name}\" for struct \"{}\" \
+                                                    in reassignment.",
+                                            struct_decl.call_path.suffix.as_str(),
+                                        ),
+                                            idx_name.span(),
+                                        ))
+                                    }
+                                    Some((field_idx, field_type_id)) => {
+                                        cur_type_id = field_type_id;
+                                        gep_indices
+                                            .push(Constant::get_uint(context, 64, field_idx));
+                                    }
+                                }
+                            }
+                            (
+                                ProjectionKind::TupleField { index, .. },
+                                TypeInfo::Tuple(field_tys),
+                            ) => {
+                                cur_type_id = field_tys[*index].type_id;
+                                gep_indices.push(Constant::get_uint(context, 64, *index as u64));
                             }
-                            Some((field_idx, field_type_id)) => {
-                                cur_type_id = field_type_id;
-                                gep_indices.push(Constant::get_uint(context, 64, field_idx));
+                            (
+                                ProjectionKind::ArrayIndex { index, .. },
+                                TypeInfo::Array(elem_ty, _),
+                            ) => {
+                                cur_type_id = elem_ty.type_id;
+                                let val = return_on_termination_or_extract!(
+                                    self.compile_expression_to_value(context, md_mgr, index)?
+                                );
```

Comment:
Excellent point! I didn't think about it at all and just took over the existing behavior. Thinking of it, just my personal view, I see LHS -> RHS to be more _logical_, but RHS -> LHS to be more _pshyological_ so to say :smile: The latter fits better in the mental model of variable declarations, I think that's why.

Out of curiosity, I've checked it for Rust, C#, and GCC (I hope other C++ compilers behave the same :smile:):
Rust: RHS -> LHS
C#, GCC:  LHS -> RHS

Since we borrow mainly from Rust for most of the semantics Jibril and I discussed it and opted for staying with the current, Rust-like behavior. A test is added to assert it, and also an explanation in the documentation.

Reply:
For C++ (and C) the language spec actually does not specify an evaluation order, and the compiler is free to do whatever it wants. I have seen at least two examples where an evaluation order bug showed up in release mode but not in debug mode - they are hell to debug, so I tend to look out for things like that. :-)